FileProvider Security Issues

API 16

App crashes due to File Provider security issue:-

```
02-25 16:49:46.281 1371-1371/? E/AndroidRuntime: FATAL EXCEPTION: main
                                                 java.lang.IllegalStateException: Could not execute method of the activity
                                                     at android.view.View$1.onClick(View.java:3591)
                                                     at android.view.View.performClick(View.java:4084)
                                                     at android.view.View$PerformClick.run(View.java:16966)
                                                     at android.os.Handler.handleCallback(Handler.java:615)
                                                     at android.os.Handler.dispatchMessage(Handler.java:92)
                                                     at android.os.Looper.loop(Looper.java:137)
                                                     at android.app.ActivityThread.main(ActivityThread.java:4745)
                                                     at java.lang.reflect.Method.invokeNative(Native Method)
                                                     at java.lang.reflect.Method.invoke(Method.java:511)
                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
                                                     at dalvik.system.NativeStart.main(Native Method)
                                                  Caused by: java.lang.reflect.InvocationTargetException
                                                     at java.lang.reflect.Method.invokeNative(Native Method)
                                                     at java.lang.reflect.Method.invoke(Method.java:511)
                                                     at android.view.View$1.onClick(View.java:3586)
                                                     at android.view.View.performClick(View.java:4084) 
                                                     at android.view.View$PerformClick.run(View.java:16966) 
                                                     at android.os.Handler.handleCallback(Handler.java:615) 
                                                     at android.os.Handler.dispatchMessage(Handler.java:92) 
                                                     at android.os.Looper.loop(Looper.java:137) 
                                                     at android.app.ActivityThread.main(ActivityThread.java:4745) 
                                                     at java.lang.reflect.Method.invokeNative(Native Method) 
                                                     at java.lang.reflect.Method.invoke(Method.java:511) 
                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786) 
                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553) 
                                                     at dalvik.system.NativeStart.main(Native Method) 
                                                  Caused by: java.lang.SecurityException: Permission Denial: opening provider android.support.v4.content.FileProvider from ProcessRecord{537e1d30 1371:com.android.camera/u0a22} (pid=1371, uid=10022) that is not exported from uid 10048
                                                     at android.os.Parcel.readException(Parcel.java:1425)
                                                     at android.os.Parcel.readException(Parcel.java:1379)
                                                     at android.app.ActivityManagerProxy.getContentProvider(ActivityManagerNative.java:2354)
                                                     at android.app.ActivityThread.acquireProvider(ActivityThread.java:4219)
                                                     at android.app.ContextImpl$ApplicationContentResolver.acquireUnstableProvider(ContextImpl.java:1711)
                                                     at android.content.ContentResolver.acquireUnstableProvider(ContentResolver.java:1099)
                                                     at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:614)
                                                     at android.content.ContentResolver.openOutputStream(ContentResolver.java:487)
                                                     at android.content.ContentResolver.openOutputStream(ContentResolver.java:463)
                                                     at com.android.camera.Camera.doAttach(Camera.java:1385)
                                                     at com.android.camera.Camera.onReviewDoneClicked(Camera.java:1362)
                                                     at java.lang.reflect.Method.invokeNative(Native Method) 
                                                     at java.lang.reflect.Method.invoke(Method.java:511) 
                                                     at android.view.View$1.onClick(View.java:3586) 
                                                     at android.view.View.performClick(View.java:4084) 
                                                     at android.view.View$PerformClick.run(View.java:16966) 
                                                     at android.os.Handler.handleCallback(Handler.java:615) 
                                                     at android.os.Handler.dispatchMessage(Handler.java:92) 
                                                     at android.os.Looper.loop(Looper.java:137) 
                                                     at android.app.ActivityThread.main(ActivityThread.java:4745) 
                                                     at java.lang.reflect.Method.invokeNative(Native Method) 
                                                     at java.lang.reflect.Method.invoke(Method.java:511) 
                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786) 
                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553) 
                                                     at dalvik.system.NativeStart.main(Native Method) 

```

--- a/paracamera/src/main/java/com/mindorks/paracamera/Camera.java
+++ b/paracamera/src/main/java/com/mindorks/paracamera/Camera.java
@@ -4,12 +4,16 @@ import android.app.Activity;
 import android.app.Fragment;
 import android.content.Context;
 import android.content.Intent;
+import android.content.pm.PackageManager;
+import android.content.pm.ResolveInfo;
 import android.graphics.Bitmap;
+import android.net.Uri;
 import android.provider.MediaStore;
 import android.support.v4.content.FileProvider;
 import android.text.TextUtils;
 
 import java.io.File;
+import java.util.List;
 
 /**
  * Created by ali on 4/11/15.
@@ -72,6 +76,28 @@ public class Camera {
         authority = context.getApplicationContext().getPackageName() + ".imageprovider";
     }
 
+    private void setUpIntent(Intent takePictureIntent){
+        File photoFile = Utils.createImageFile(context, dirName, imageName, imageType);
+        if (photoFile != null) {
+
+            cameraBitmapPath = photoFile.getAbsolutePath();
+
+            Uri uri=FileProvider.getUriForFile(context, authority, photoFile);
+
+            takePictureIntent.putExtra(
+                    MediaStore.EXTRA_OUTPUT,
+                    uri);
+            List<ResolveInfo> resInfoList = context.getPackageManager().queryIntentActivities(takePictureIntent, PackageManager.MATCH_DEFAULT_ONLY);
+            for (ResolveInfo resolveInfo : resInfoList) {
+                String packageName = resolveInfo.activityInfo.packageName;
+                context.grantUriPermission(packageName, uri, Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION);
+            }
+        }
+        else {
+            throw new NullPointerException("Image file could not be created");
+        }
+    }
+
     /**
      * Initiate the existing camera apps
      *
@@ -82,60 +108,26 @@ public class Camera {
         switch (mode) {
             case ACTIVITY:
                 if (takePictureIntent.resolveActivity(activity.getPackageManager()) != null) {
-
-                    File photoFile = Utils.createImageFile(context, dirName, imageName, imageType);
-                    if (photoFile != null) {
-
-                        cameraBitmapPath = photoFile.getAbsolutePath();
-
-                        takePictureIntent.putExtra(
-                                MediaStore.EXTRA_OUTPUT,
-                                FileProvider.getUriForFile(context, authority, photoFile));
-
-                        activity.startActivityForResult(takePictureIntent, REQUEST_TAKE_PHOTO);
-                    } else {
-                        throw new NullPointerException("Image file could not be created");
-                    }
+                    setUpIntent(takePictureIntent);
+                    activity.startActivityForResult(takePictureIntent,REQUEST_TAKE_PHOTO);
                 } else {
                     throw new IllegalAccessException("Unable to open camera");
                 }
                 break;
+
             case FRAGMENT:
                 if (takePictureIntent.resolveActivity(fragment.getActivity().getPackageManager()) != null) {
-
-                    File photoFile = Utils.createImageFile(context, dirName, imageName, imageType);
-                    if (photoFile != null) {
-
-                        cameraBitmapPath = photoFile.getAbsolutePath();
-
-                        takePictureIntent.putExtra(
-                                MediaStore.EXTRA_OUTPUT,
-                                FileProvider.getUriForFile(context, authority, photoFile));
-
-                        fragment.startActivityForResult(takePictureIntent, REQUEST_TAKE_PHOTO);
-                    } else {
-                        throw new NullPointerException("Image file could not be created");
-                    }
+                        setUpIntent(takePictureIntent);
+                        fragment.startActivityForResult(takePictureIntent,REQUEST_TAKE_PHOTO);
                 } else {
                     throw new IllegalAccessException("Unable to open camera");
                 }
                 break;
+
             case COMPAT_FRAGMENT:
                 if (takePictureIntent.resolveActivity(compatFragment.getActivity().getPackageManager()) != null) {
-
-                    File photoFile = Utils.createImageFile(context, dirName, imageName, imageType);
-                    if (photoFile != null) {
-
-                        cameraBitmapPath = photoFile.getAbsolutePath();
-
-                        takePictureIntent.putExtra(
-                                MediaStore.EXTRA_OUTPUT,
-                                FileProvider.getUriForFile(context, authority, photoFile));
-
-                        compatFragment.startActivityForResult(takePictureIntent, REQUEST_TAKE_PHOTO);
-                    } else {
-                        throw new NullPointerException("Image file could not be created");
-                    }
+                        setUpIntent(takePictureIntent);
+                        compatFragment.startActivityForResult(takePictureIntent,REQUEST_TAKE_PHOTO);
                 } else {
                     throw new IllegalAccessException("Unable to open camera");
                 }

